File: 2025/JuyoungMoon/week12/2025-08-24-course-schedule.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 12: Graph (BFS/DFS/Union)
// https://github.com/juyomo/leetcode-study

// LeetCode #207.
// https://leetcode.com/problems/course-schedule/

class Solution {
public:
    bool detectCycle(const vector<vector<int>>& courses,
                     unordered_set<int>& needToTake,
                     unordered_set<int>& seenInCurrentScan,
                     int startNode) {
        if (seenInCurrentScan.count(startNode) > 0) {
            return true;
        }
        
        if (needToTake.count(startNode) == 0) {
            return false;
        }

        needToTake.erase(startNode);
        seenInCurrentScan.insert(startNode);

        for (int nb : courses[startNode]) {
            if (detectCycle(courses, needToTake, seenInCurrentScan, nb)) {
                return true;
            }
        }

        seenInCurrentScan.erase(startNode);
        return false;
    }

    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // detect cycles
        vector<vector<int>> courses(numCourses, vector<int>());

        for (const auto& pair : prerequisites) {
            int from = pair[0];
            int to = pair[1];

            courses[from].push_back(to);
        }

        unordered_set<int> needToTake;
        for (int i = 0; i < numCourses; i++) {
            needToTake.insert(i);
        }
        
        while (!needToTake.empty()) {
            // pick a random element in needToTake
            int startNode = 0;
            for (int n : needToTake) {
                startNode = n;
                break;
            }
            unordered_set<int> seenInCurrentScan;
            
            if (detectCycle(courses, needToTake, seenInCurrentScan, startNode)) {
                return false;
            }

            /*tovisit.push(startNode);

            while (tovisit.size() > 0) {
                int curr = tovisit.top();
                seenInCurrentScan.insert(curr);
                tovisit.pop();
                needToTake.erase(curr);

                for (const int nb : courses[curr]) {
                    if (seenInCurrentScan.count(nb) > 0) {
                        return false;
                    } else {
                        tovisit.push(nb);
                    }
                }
            }*/
        }

        return true;
    }
};

File: 2025/Jinah/week12/2025-08-24-course-schedule.py
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # map each course to prereq list
        preMap = { i: [] for i in range(numCourses) }

        for crs, pre in prerequisites:
            preMap[crs].append(pre)

        # visitSet = all courses along the curr DFS path
        visitSet = set()

        def dfs(crs):
            if crs in visitSet:
                return False
            if preMap[crs] == []:
                return True

            visitSet.add(crs)

            for pre in preMap[crs]:
                if not dfs(pre):
                    return False

            visitSet.remove(crs)
            preMap[crs] = []            
            return True
        
        for crs in range(numCourses):
            if not dfs(crs):
                return False
        return True